src/resources/ble_darwin.mm

```diff
@@ -0,0 +1,1297 @@
+  if (discovered_services != nil) {
+    Array* service_classes = process->object_heap()->allocate_array(
+        static_cast<int>([discovered_services count]),
+        process->program()->null_object());
+
+    for (int i = 0; i < [discovered_services count]; i++) {
+      String* uuid = process->allocate_string([[discovered_services[i] UUIDString] UTF8String]);
+      if (uuid == null) {
+        free(peripheral);
+        ALLOCATION_FAILED;
+      }
+      service_classes->at_put(i, uuid);
+    }
+    array->at_put(3, service_classes);
+  }
+
+  NSData* manufacturer_data = peripheral->manufacturer_data();
+  if (manufacturer_data != nil) {
+    ByteArray* custom_data = process->object_heap()->allocate_internal_byte_array(
+        static_cast<int>([manufacturer_data length]));
+    ByteArray::Bytes custom_data_bytes(custom_data);
+    if (!custom_data) ALLOCATION_FAILED;
+    memcpy(custom_data_bytes.address(), manufacturer_data.bytes, [manufacturer_data length]);
+
+    array->at_put(4, custom_data);
+  }
+
+  NSNumber* is_connectable = peripheral->connectable();
+  array->at_put(5, (is_connectable != nil && is_connectable.boolValue == YES)
+                   ? process->program()->true_object()
+                   : process->program()->false_object());
+
+  delete peripheral;
+
+  return array;
+}
+
+PRIMITIVE(scan_stop) {
+  ARGS(BLECentralManagerResource, central_manager);
+  Locker locker(central_manager->scan_mutex());
+
+  if (central_manager->scan_active()) {
+    OS::signal(central_manager->stop_scan_condition());
+  }
+
+  return process->program()->null_object();
+}
+
+PRIMITIVE(connect) {
+  ARGS(BLECentralManagerResource, central_manager, Blob, address);
+
+  NSUUID* uuid = [[NSUUID alloc] initWithUUIDString:ns_string_from_blob(address)];
+
+  CBPeripheral* peripheral = central_manager->get_peripheral(uuid);
+  if (peripheral == nil) INVALID_ARGUMENT;
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  auto device =
+      _new BLERemoteDeviceResource(
+          central_manager->group(),
+          central_manager->central_manager(),
+          peripheral);
+
+  central_manager->group()->register_resource(device);
+  proxy->set_external_address(device);
+
+  peripheral.delegate = [[[TOITPeripheralDelegate alloc] initWithDevice:device] retain];
+  [central_manager->central_manager() connectPeripheral:peripheral options:nil];
+
+  return proxy;
+}
+
+PRIMITIVE(disconnect) {
+  ARGS(BLERemoteDeviceResource, device);
+
+  [device->central_manager() cancelPeripheralConnection:device->peripheral()];
+
+  return process->program()->null_object();
+}
+
+PRIMITIVE(release_resource) {
+  ARGS(Resource, resource);
+
+  resource->resource_group()->unregister_resource(resource);
+
+  return process->program()->null_object();
+}
+
+PRIMITIVE(discover_services) {
+  ARGS(BLERemoteDeviceResource, device, Array, raw_service_uuids);
+
+  Error* err;
```

Comment:
I know the `ns_uuid_array_from_array_of_strings` function does it for us, but it doesn't really cost to set it here.
```suggestion
  Error* err = null;
```